testhelper/convenience.go

```diff
@@ -167,15 +167,13 @@ func deepDiffEqual(expected, actual reflect.Value, visited map[visit]bool, path
 
 		for _, k := range keys {
 			expectedValue := expected.MapIndex(k)
-			actualValue := expected.MapIndex(k)
+			actualValue := actual.MapIndex(k)
 
 			if !expectedValue.IsValid() {
-				logDifference(path, nil, actual.Interface())
-				return
```

Comment:
I spent several hours trying to sort this one out. What I have in the PR is not correct. Similarly, my "What about..." idea is incorrect, too.

The problem is when a map contains a nested map. The nested map will be turned into an initiated map and return `false` for `IsNil()`.

For example:

```
--- FAIL: TestBuildRequestBody (0.00s)
        convenience.go:251: Failure in params_test.go, line 202:at [auth] expected <nil>, but got map[string]interface {}{"token":map[string]interface {}{}, "or_fields":map[string]interface {}{}, "passwordCredentials":map[string]interface {}{"username":"me", "password":"s
wordfish"}}
        convenience.go:35: Failure in params_test.go, line 202: The structures were different.
```

The above is being triggered [here](https://github.com/jtopjian/gophercloud/blob/ea65cfab418dcd5b5e245ef4c28eff74641ef104/testhelper/convenience.go#L168) when trying to test against [this](https://github.com/jtopjian/gophercloud/blob/ea65cfab418dcd5b5e245ef4c28eff74641ef104/testing/params_test.go#L167-L182), specifically when the `for` loop hits [`token`](https://github.com/jtopjian/gophercloud/blob/ea65cfab418dcd5b5e245ef4c28eff74641ef104/testing/params_test.go#L158).

`expected` is set to "Invalid" but "actual" has `"token":map[string]interface {}{}`.

I tried a different route by changing the [struct](https://github.com/jtopjian/gophercloud/blob/ea65cfab418dcd5b5e245ef4c28eff74641ef104/testing/params_test.go#L146-L161) in the test to this:

```go
type AuthOptions struct {
		*PasswordCredentials `json:"passwordCredentials,omitempty" xor:"TokenCredentials"`

		// The TenantID and TenantName fields are optional for the Identity V2 API.
		// Some providers allow you to specify a TenantName instead of the TenantId.
		// Some require both. Your provider's authentication policies will determine
		// how these fields influence authentication.
		TenantID   string `json:"tenantId,omitempty"`
		TenantName string `json:"tenantName,omitempty"`

		// TokenCredentials allows users to authenticate (possibly as another user) with an
		// authentication token ID.
		*TokenCredentials `json:"token,omitempty" xor:"PasswordCredentials"`

		OrFields *orFields `json:"or_fields,omitempty"`
	}
```

Which made the test pass, but then I run into an error here:

```
$ go test -v openstack/cdn/v1/base/testing/*
=== RUN   TestGetHomeDocument
--- FAIL: TestGetHomeDocument (0.00s)
        convenience.go:256: Failure in requests_test.go, line 34: at  expected base.HomeDocument{"rel/cdn":map[string]interface {}{"href-template":"services{?marker,limit}", "href-vars":map[string]interface {}{"marker":"param/marker", "limit":"param/limit"}, "hints":map[string]interface {}{"allow":[]string{"GET"}, "formats":map[string]interface {}{"application/json":interface {}(nil)}}}}, but got <nil>
```

Would it be best to close this PR and break the two points of discussion (`RemainingKeys` and this part of the code) into separate issues? Or just focus on one for the moment.